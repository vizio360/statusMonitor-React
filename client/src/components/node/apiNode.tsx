import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {Node, NodeStatus, INodeEvents} from './';
import {EndpointOptions, Defaults, jsPlumb, jsPlumbInstance} from 'jsPlumb';
import apiImage from '@images/api.png';
import {Fetch, IStatus, IService} from '@app/fetchData';

interface IAPINode {
  jsPlumb: jsPlumbInstance;
  service: IService;
  events: INodeEvents;
}

class APINode extends React.Component<IAPINode> {
  render() {
    return (
      <Node
        id={this.props.service.id}
        name={this.props.service.name}
        jsPlumb={this.props.jsPlumb}
        image={apiImage}
        status={NodeStatus.HEALTHY}
        events={this.props.events}
        uiProps={this.props.service.uiProps}
      />
    );
  }
}

export default APINode;
