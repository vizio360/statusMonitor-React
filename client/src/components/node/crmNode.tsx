import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {INodeEvents, Node, NodeStatus} from './';
import {EndpointOptions, Defaults, jsPlumb, jsPlumbInstance} from 'jsPlumb';
import crmImage from '@images/crm.png';
import {Fetch, IStatus, IService} from '@app/fetchData';

interface ICRMNode {
  jsPlumb: jsPlumbInstance;
  service: IService;
  events?: INodeEvents;
}

class CRMNode extends React.Component<ICRMNode> {
  constructor(props: ICRMNode) {
    super(props);
  }
  componentDidMount() {
    console.log('mounted!');
  }

  componentWillUnmount() {
    console.log('umounted');
  }

  componentWillReceiveProps(props: ICRMNode) {
    console.log(props);
  }

  render() {
    return (
      <Fetch service={this.props.service}>
        {(status: IStatus) => (
          <Node
            id={this.props.service.id}
            name={this.props.service.name}
            jsPlumb={this.props.jsPlumb}
            image={crmImage}
            status={
              status.status === 'Unhealthy'
                ? NodeStatus.UNHEALTHY
                : NodeStatus.HEALTHY
            }
            events={this.props.events}
            uiProps={this.props.service.uiProps}
          />
        )}
      </Fetch>
    );
  }
}

export default CRMNode;
