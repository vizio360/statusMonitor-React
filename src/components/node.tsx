import * as React from 'react';
import * as ReactDOM from 'react-dom';

interface INodeProps {
  image: Element;
  name: string;
}

class Node extends React.Component<INodeProps> {}

export default Node;
