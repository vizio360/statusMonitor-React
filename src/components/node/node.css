.node {
  width: 100px;
  height: 100px;
  position: absolute;
  text-align: center;
}
.nodeOverlay {
  top: 0;
  left: 0;
  width: 100px;
  height: 100px;
  position: absolute;
}
