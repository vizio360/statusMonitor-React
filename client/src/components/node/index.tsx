import * as React from 'react';
import * as ReactDOM from 'react-dom';
import './node.css';
import fireImage from '@images/fire.gif';
import {
  Status,
  ServiceType,
  IService,
  IServiceLastKnownState,
} from '@dataTypes';

import crmImage from '@images/crm.png';
import dbImage from '@images/db.png';
import apiImage from '@images/api.png';

interface INodeEvents {
  onDoubleClick: (id: string) => void;
}

interface INodeProps {
  service: IService;
  serviceState: IServiceLastKnownState;
  className?: string;
  events?: INodeEvents;
}

class Node extends React.Component<INodeProps> {
  nodeEvents: INodeEvents;

  constructor(props: INodeProps) {
    super(props);
    this.nodeEvents = props.events;
    this.onDoubleClick = this.onDoubleClick.bind(this);
  }

  onDoubleClick(e: React.MouseEvent<HTMLElement>) {
    if (this.nodeEvents && this.nodeEvents.onDoubleClick)
      this.nodeEvents.onDoubleClick(this.props.service.id);
  }

  getNodeTypeImage(serviceType: ServiceType) {
    let img: string;
    switch (serviceType) {
      case ServiceType.DB:
        img = dbImage;
        break;
      case ServiceType.API:
        img = apiImage;
        break;
      case ServiceType.CRM:
        img = crmImage;
        break;
    }
    return img;
  }

  render() {
    return (
      <div
        id={this.props.service.id}
        className={this.props.className ? this.props.className : 'node'}
        style={this.props.service.uiProps}
        onDoubleClick={this.onDoubleClick}>
        <img
          src={this.getNodeTypeImage(this.props.service.type)}
          width="100%"
          height="100%"
        />
        {this.props.serviceState.status !== Status.HEALTHY ? (
          <img
            className="nodeOverlay"
            src={fireImage}
            width="100%"
            height="100%"
          />
        ) : (
          ''
        )}
        <p>{this.props.service.name}</p>
      </div>
    );
  }
}

export {Node, INodeEvents};
