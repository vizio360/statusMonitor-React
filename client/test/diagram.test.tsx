import React from 'react';
import {Node} from '@app/components/node';
import NavBar from '@app/navbar';
import NodeEditor from '@app/nodeEditor';

//ignoring typescript errors as the jsPlumb mocks
//export more symbols than the original jsPlumb
//these symbols are used to have quick access
//to all jsPlumb function mocks
//MOCKS are in __mocks__/jsPlumb
import {
  jsPlumb,
  //@ts-ignore
  draggable,
  //@ts-ignore
  addEndpoint,
  //@ts-ignore
  setContainer,
  //@ts-ignore
  getInstance,
  //@ts-ignore
  bind,
  //@ts-ignore
  reset,
  //@ts-ignore
  setDraggable,
} from 'jsPlumb';

import Diagram from '@app/diagram';
import renderer from 'react-test-renderer';
import servicesMock from '@mocks/services.json';
import connectionsMock from '@mocks/connections.json';
import lastKnownStatesMock from '@mocks/lastKnownStates.json';
import {IStatusMonitorClient} from '@app/wsClient';
import {
  ServiceType,
  IService,
  Status,
  IServiceLastKnownState,
} from '@dataTypes';

describe('Diagram', () => {
  const clone = (array: any) => {
    return JSON.parse(JSON.stringify(array));
  };
  const mockClient: any = {
    connect: jest.fn().mockResolvedValue(''),
    getServices: jest.fn().mockImplementation(() => {
      return clone(servicesMock);
    }),
    getConnections: jest.fn().mockImplementation(() => {
      return clone(connectionsMock);
    }),
    getServicesLastKnownState: jest.fn().mockImplementation(() => {
      return clone(lastKnownStatesMock);
    }),
    onUpdate: jest.fn().mockImplementation(cb => {}),
    onReload: jest.fn().mockImplementation(cb => {}),
    onError: jest.fn().mockImplementation(cb => {}),
    onDisconnect: jest.fn().mockImplementation(cb => {}),
  };

  let createDiagramComponent = async () => {
    const cdmSpy = jest.spyOn(Diagram.prototype, 'componentDidMount');
    let d = renderer.create(<Diagram wsClient={mockClient} />);
    await cdmSpy.mock.results[0].value;
    return d;
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  const checkJSPlumbInit = (serv: IService[]) => {
    expect(getInstance).toHaveBeenCalledWith({
      Connector: 'StateMachine',
      Anchors: ['Left', 'BottomRight'],
    });
    expect(setContainer).toHaveBeenCalledWith('jsPlumbDiagram');
    expect(bind).toHaveBeenNthCalledWith(1, 'connection', expect.any(Function));
    expect(bind).toHaveBeenNthCalledWith(
      2,
      'connectionDetached',
      expect.any(Function),
    );
    expect(addEndpoint.mock.calls.length).toBe(servicesMock.length * 2);
    expect(draggable.mock.calls.length).toBe(servicesMock.length);
    expect(setDraggable.mock.calls.length).toBe(servicesMock.length);
  };

  it('can be instatiated', async () => {
    let diagram = await createDiagramComponent();
    let diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
    expect(mockClient.connect).toHaveBeenCalledWith(
      'ws://localhost:3333/channel',
    );
    expect(mockClient.getServices).toHaveBeenCalled();
    expect(mockClient.getConnections).toHaveBeenCalled();
    expect(mockClient.getServicesLastKnownState).toHaveBeenCalled();

    checkJSPlumbInit(servicesMock as IService[]);
  });

  it('updates when an update state is received', async () => {
    let diagram = await createDiagramComponent();
    let diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
    const updateCallback = mockClient.onUpdate.mock.calls[0][0];
    const state: IServiceLastKnownState = {
      serviceId: '1',
      status: Status.UNHEALTHY,
      responseBody: 'casjcnaksd',
    };
    updateCallback(state);
    diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
  });

  it('reloads when a reload message is received', async () => {
    let diagram = await createDiagramComponent();
    let diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
    const reloadCallback = mockClient.onReload.mock.calls[0][0];

    mockClient.getServices.mockImplementationOnce(() => {
      return clone(servicesMock.slice(2));
    });
    mockClient.getConnections.mockImplementationOnce(() => {
      return clone(connectionsMock.slice(2));
    });
    mockClient.getServicesLastKnownState.mockImplementationOnce(() => {
      return clone(lastKnownStatesMock.slice(2));
    });

    reloadCallback();
    expect(reset).toHaveBeenCalled();
    diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
  });

  it('resets itself if cancelling an edit', async () => {
    let diagram = await createDiagramComponent();
    let component = diagram.root;
    let initSpy = jest.spyOn(Diagram.prototype, 'init');

    let navBar = component.findByType(NavBar);
    navBar.props.onEdit();
    navBar.props.onCancel();
    expect(initSpy).toHaveBeenCalled();
    let diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
  });

  it('adds a new node on node editor confirmation', async () => {
    let diagram = await createDiagramComponent();
    let diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
    let component = diagram.root;

    let newService: IService = {
      id: '',
      name: 'newService',
      uri: 'newService URI',
      timeout: 30,
      categories: ['cat1'],
      type: ServiceType.CRM,
      uiProps: {
        top: 10,
        left: 15,
      },
    };
    let nodeEditor = component.findByType(NodeEditor);
    nodeEditor.props.onConfirm(newService);
    diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
  });

  it('cancels the node editor modal', async () => {
    let diagram = await createDiagramComponent();
    let component = diagram.root;
    let node = component.findAllByType(Node)[0];
    node.props.events.onDoubleClick(node.props.service.id);
    let diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
    let nodeEditor = component.findByType(NodeEditor);
    nodeEditor.props.onCancel();
    diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
  });

  it('deletes a node', async () => {
    let diagram = await createDiagramComponent();
    let component = diagram.root;
    let diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();

    let node = component.findAllByType(Node)[1];
    node.props.events.onDoubleClick(node.props.service.id);

    let nodeEditor = component.findByType(NodeEditor);
    nodeEditor.props.onDelete(nodeEditor.props.node);

    diagramJson = diagram.toJSON();
    expect(diagramJson).toMatchSnapshot();
  });

  it('sends the config to the server on save', async () => {
    fetchMock.mockResponses(['', {status: 200}], ['', {status: 200}]);
    let diagram = await createDiagramComponent();
    let component = diagram.root;
    let navBar = component.findByType(NavBar);
    navBar.props.onSave();
    //NOT IDEAL but only way I could test a timeout after a fetch
    const timeout = () => {
      return new Promise(resolve => setTimeout(resolve, 600));
    };
    await timeout();
    expect(fetchMock.mock.calls.length).toEqual(2);
    expect(fetchMock.mock.calls[0][0]).toEqual('/config');
    expect(fetchMock.mock.calls[1][0]).toEqual('/reload');
  });
});
