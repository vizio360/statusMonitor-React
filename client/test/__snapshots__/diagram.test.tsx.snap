// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagram can be instatiated 1`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="#"
    >
      Monitor
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        className="navbar-nav mr-auto"
      >
        <li
          className="nav-item active"
        >
          <button
            className="btn btn-primary mr-sm-2"
            onClick={[Function]}
            type="button"
          >
            Add Node
          </button>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="btn btn-primary mr-sm-2"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    id="hello"
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    <div
      className="node"
      id="1"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 7,
          "top": 68,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        serviceTest
      </p>
    </div>
    <div
      className="node"
      id="2"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 712,
          "top": 441,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        CRM
      </p>
    </div>
    <div
      className="node"
      id="3"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 218,
          "top": 75,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        DataHub
      </p>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalLabel"
    className="modal fade"
    id="nodeEditor"
    role="dialog"
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
            id="exampleModalLabel"
          >
            Add Node
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <form>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="name"
              >
                Name:
              </label>
              <input
                className="form-control"
                id="name"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="uri"
              >
                Uri:
              </label>
              <input
                className="form-control"
                id="uri"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="timeout"
              >
                Timeout:
              </label>
              <input
                className="form-control"
                id="timeout"
                onChange={[Function]}
                type="text"
                value={20}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="categories"
              >
                Categories:
              </label>
              <input
                className="form-control"
                id="categories"
                onChange={[Function]}
                type="text"
                value={Array []}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="type"
              >
                Type:
              </label>
              <select
                className="form-control"
                id="type"
                onChange={[Function]}
                value="EMPTY"
              >
                <option
                  value="CRM"
                >
                  CRM
                </option>
                <option
                  value="DB"
                >
                  DB
                </option>
                <option
                  value="API"
                >
                  API
                </option>
              </select>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-dismiss="modal"
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            Ok
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Diagram reloads when a reload message is received 1`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="#"
    >
      Monitor
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        className="navbar-nav mr-auto"
      >
        <li
          className="nav-item active"
        >
          <button
            className="btn btn-primary mr-sm-2"
            onClick={[Function]}
            type="button"
          >
            Add Node
          </button>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="btn btn-primary mr-sm-2"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    id="hello"
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    <div
      className="node"
      id="1"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 7,
          "top": 68,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        serviceTest
      </p>
    </div>
    <div
      className="node"
      id="2"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 712,
          "top": 441,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        CRM
      </p>
    </div>
    <div
      className="node"
      id="3"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 218,
          "top": 75,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        DataHub
      </p>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalLabel"
    className="modal fade"
    id="nodeEditor"
    role="dialog"
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
            id="exampleModalLabel"
          >
            Add Node
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <form>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="name"
              >
                Name:
              </label>
              <input
                className="form-control"
                id="name"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="uri"
              >
                Uri:
              </label>
              <input
                className="form-control"
                id="uri"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="timeout"
              >
                Timeout:
              </label>
              <input
                className="form-control"
                id="timeout"
                onChange={[Function]}
                type="text"
                value={20}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="categories"
              >
                Categories:
              </label>
              <input
                className="form-control"
                id="categories"
                onChange={[Function]}
                type="text"
                value={Array []}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="type"
              >
                Type:
              </label>
              <select
                className="form-control"
                id="type"
                onChange={[Function]}
                value="EMPTY"
              >
                <option
                  value="CRM"
                >
                  CRM
                </option>
                <option
                  value="DB"
                >
                  DB
                </option>
                <option
                  value="API"
                >
                  API
                </option>
              </select>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-dismiss="modal"
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            Ok
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Diagram reloads when a reload message is received 2`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="#"
    >
      Monitor
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        className="navbar-nav mr-auto"
      >
        <li
          className="nav-item active"
        >
          <button
            className="btn btn-primary mr-sm-2"
            onClick={[Function]}
            type="button"
          >
            Add Node
          </button>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="btn btn-primary mr-sm-2"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    id="hello"
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    <div
      className="node"
      id="3"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 218,
          "top": 75,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        DataHub
      </p>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalLabel"
    className="modal fade"
    id="nodeEditor"
    role="dialog"
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
            id="exampleModalLabel"
          >
            Add Node
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <form>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="name"
              >
                Name:
              </label>
              <input
                className="form-control"
                id="name"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="uri"
              >
                Uri:
              </label>
              <input
                className="form-control"
                id="uri"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="timeout"
              >
                Timeout:
              </label>
              <input
                className="form-control"
                id="timeout"
                onChange={[Function]}
                type="text"
                value={20}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="categories"
              >
                Categories:
              </label>
              <input
                className="form-control"
                id="categories"
                onChange={[Function]}
                type="text"
                value={Array []}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="type"
              >
                Type:
              </label>
              <select
                className="form-control"
                id="type"
                onChange={[Function]}
                value="EMPTY"
              >
                <option
                  value="CRM"
                >
                  CRM
                </option>
                <option
                  value="DB"
                >
                  DB
                </option>
                <option
                  value="API"
                >
                  API
                </option>
              </select>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-dismiss="modal"
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            Ok
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Diagram updates when an update state is received 1`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="#"
    >
      Monitor
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        className="navbar-nav mr-auto"
      >
        <li
          className="nav-item active"
        >
          <button
            className="btn btn-primary mr-sm-2"
            onClick={[Function]}
            type="button"
          >
            Add Node
          </button>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="btn btn-primary mr-sm-2"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    id="hello"
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    <div
      className="node"
      id="1"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 7,
          "top": 68,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        serviceTest
      </p>
    </div>
    <div
      className="node"
      id="2"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 712,
          "top": 441,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        CRM
      </p>
    </div>
    <div
      className="node"
      id="3"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 218,
          "top": 75,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        DataHub
      </p>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalLabel"
    className="modal fade"
    id="nodeEditor"
    role="dialog"
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
            id="exampleModalLabel"
          >
            Add Node
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <form>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="name"
              >
                Name:
              </label>
              <input
                className="form-control"
                id="name"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="uri"
              >
                Uri:
              </label>
              <input
                className="form-control"
                id="uri"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="timeout"
              >
                Timeout:
              </label>
              <input
                className="form-control"
                id="timeout"
                onChange={[Function]}
                type="text"
                value={20}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="categories"
              >
                Categories:
              </label>
              <input
                className="form-control"
                id="categories"
                onChange={[Function]}
                type="text"
                value={Array []}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="type"
              >
                Type:
              </label>
              <select
                className="form-control"
                id="type"
                onChange={[Function]}
                value="EMPTY"
              >
                <option
                  value="CRM"
                >
                  CRM
                </option>
                <option
                  value="DB"
                >
                  DB
                </option>
                <option
                  value="API"
                >
                  API
                </option>
              </select>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-dismiss="modal"
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            Ok
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Diagram updates when an update state is received 2`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="#"
    >
      Monitor
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        className="navbar-nav mr-auto"
      >
        <li
          className="nav-item active"
        >
          <button
            className="btn btn-primary mr-sm-2"
            onClick={[Function]}
            type="button"
          >
            Add Node
          </button>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="btn btn-primary mr-sm-2"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    id="hello"
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    <div
      className="node"
      id="1"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 7,
          "top": 68,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      <img
        className="nodeOverlay"
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      <p>
        serviceTest
      </p>
    </div>
    <div
      className="node"
      id="2"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 712,
          "top": 441,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        CRM
      </p>
    </div>
    <div
      className="node"
      id="3"
      onDoubleClick={[Function]}
      style={
        Object {
          "left": 218,
          "top": 75,
        }
      }
    >
      <img
        height="100%"
        src="test-file-stub"
        width="100%"
      />
      
      <p>
        DataHub
      </p>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalLabel"
    className="modal fade"
    id="nodeEditor"
    role="dialog"
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
            id="exampleModalLabel"
          >
            Add Node
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <form>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="name"
              >
                Name:
              </label>
              <input
                className="form-control"
                id="name"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="uri"
              >
                Uri:
              </label>
              <input
                className="form-control"
                id="uri"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="timeout"
              >
                Timeout:
              </label>
              <input
                className="form-control"
                id="timeout"
                onChange={[Function]}
                type="text"
                value={20}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="categories"
              >
                Categories:
              </label>
              <input
                className="form-control"
                id="categories"
                onChange={[Function]}
                type="text"
                value={Array []}
              />
            </div>
            <div
              className="form-group"
            >
              <label
                className="col-form-label"
                htmlFor="type"
              >
                Type:
              </label>
              <select
                className="form-control"
                id="type"
                onChange={[Function]}
                value="EMPTY"
              >
                <option
                  value="CRM"
                >
                  CRM
                </option>
                <option
                  value="DB"
                >
                  DB
                </option>
                <option
                  value="API"
                >
                  API
                </option>
              </select>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-dismiss="modal"
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            Ok
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
